drivers/counter/counter_ifx_cat1.c

```diff
@@ -0,0 +1,537 @@
+{
+	__ASSERT_NO_MSG(dev != NULL);
+
+	struct ifx_cat1_counter_data *const data = dev->data;
+
+	return data->counter_cfg.period;
+}
+
+static inline bool counter_is_bit_mask(uint32_t val)
+{
+	/* Return true if value equals 2^n - 1 */
+	return !(val & (val + 1U));
+}
+
+static uint32_t ifx_cat1_counter_ticks_add(uint32_t val1, uint32_t val2, uint32_t top)
+{
+	uint32_t to_top;
+
+	/* refer to https://tbrindus.ca/how-builtin-expect-works/ for 'likely' usage */
+	if (likely(counter_is_bit_mask(top))) {
+		return (val1 + val2) & top;
+	}
+
+	to_top = top - val1;
+
+	return (val2 <= to_top) ? (val1 + val2) : (val2 - to_top - 1U);
+}
+
+static uint32_t ifx_cat1_counter_ticks_sub(uint32_t val, uint32_t old, uint32_t top)
+{
+	/* refer to https://tbrindus.ca/how-builtin-expect-works/ for 'likely' usage */
+	if (likely(counter_is_bit_mask(top))) {
+		return (val - old) & top;
+	}
+
+	/* if top is not 2^n-1 */
+	return (val >= old) ? (val - old) : (val + top + 1U - old);
+}
+
+static int ifx_cat1_counter_set_alarm(const struct device *dev, uint8_t chan_id,
+				      const struct counter_alarm_cfg *alarm_cfg)
+{
+	ARG_UNUSED(chan_id);
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(alarm_cfg != NULL);
+
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+
+	uint32_t val = alarm_cfg->ticks;
+	uint32_t top_val = ifx_cat1_counter_get_top_value(dev);
+	uint32_t flags = alarm_cfg->flags;
+	uint32_t max_rel_val;
+	bool absolute = ((flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) ? false : true;
+	bool irq_on_late;
+	int err = 0;
+
+	/* Checks if compare value is not less then period value */
+	if (alarm_cfg->ticks > top_val) {
+		return -EINVAL;
+	}
+
+	if (absolute) {
+		max_rel_val = top_val - data->guard_period;
+		irq_on_late = ((flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE) == 0) ? false : true;
+	} else {
+		/* If relative value is smaller than half of the counter range it is assumed
+		 * that there is a risk of setting value too late and late detection algorithm
+		 * must be applied. When late setting is detected, interrupt shall be
+		 * triggered for immediate expiration of the timer. Detection is performed
+		 * by limiting relative distance between CC and counter.
+		 *
+		 * Note that half of counter range is an arbitrary value.
+		 */
+		irq_on_late = val < (top_val / 2U);
+
+		/* limit max to detect short relative being set too late. */
+		max_rel_val = irq_on_late ? (top_val / 2U) : top_val;
+		val = ifx_cat1_counter_ticks_add(cyhal_timer_read(&data->counter_obj),
+						 val, top_val);
+	}
+
+	/* Decrement value to detect also case when val == counter_read(dev). Otherwise,
+	 * condition would need to include comparing diff against 0.
+	 */
+	uint32_t curr = cyhal_timer_read(&data->counter_obj);
+	uint32_t diff = ifx_cat1_counter_ticks_sub((val - 1), curr, top_val);
+
+	if ((absolute && (val < curr)) || (diff > max_rel_val)) {
+		if (absolute) {
+			err = -ETIME;
```

Comment:
The exit path of this function is a bit hard to follow, how about:
- change the function exit on 445 to just `return 0`
- move this after the `irq_on_late` check and do a `return -ETIME` directly (the function behavior is the same)
- drop `err`, you won't need it anymore

I think that would make it more obvious to understand what's going on.

Reply:
Changed in most recent push